[xhamyo] iP

xhamyo opened the pull request:
# **Hamyo** Chatbot 🤖 

> "Strive not to be a success, but rather to be of value." - Albert Einstein (1879-1955) [[Read More]](https://en.wikipedia.org/wiki/Albert_Einstein)

### Too many tasks to keep track of? **Hamyo** allows you to manage your
- 📝 ToDo
- 🕙 Deadline
- 💼 Event

### Functionalities supported by **Hamyo**
1. Add and Delete Tasks
2. Mark and Unmark Tasks
3. List all Tasks
4. List Tasks by Date or Keyword

### Features of **Hamyo**
- [X] Store tasks on Local Device
- [X] Text-Based Interface
- [ ] ~~*Graphical User Interface*~~ *(coming soon)*

```java
public class Hamyo {
    public static void main(String[] args) {
        System.out.println("USE HAMYO TODAY!");
    }
}
```

> [!IMPORTANT]
> The `Hamyo` Chatbot is still a work in progress! 😄

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

xhamyo committed:
Implement Level 0

Changed the name of chatbot from Duke to Hamyo.

Implemented skeletal version of chatbot that simply greets the user and exits.

xhamyo committed:
Implement Level 1

Chatbot echos commands entered by user, and exits when user types the command "bye".

HAMYO logo changed to a minimalistic version.

Added functions to printLine() and printLogo() to improve readability.

xhamyo committed:
Implement Level 2

Gave chatbot the ability to store whatever text entered by user via "add [item]" command.

Gave chatbot the ability to display text back to the user when requested via "list" command.

xhamyo committed:
Implement Level 3

Gave chatbot the ability to mark tasks as done.
Gave chatbot the ability to change the task status back to not done.

xhamyo committed:
Add Task class

Task class to represent a task. Possible children include ToDos, Deadlines and Events.

xhamyo committed:
Implement Level 4

Added support for tracking ToDo (no time/ date attached), Deadline (done before specific date/ time), Events (occurs between two specific date/time).

xhamyo committed:
Implement A-TextUiTesting

Added Automated Text UI Testing for simple test cases (3 tasks types) and exceptions.

xhamyo committed:
Implement Level 5

Added HamyoException to represent exceptions specific to Hamyo.

Gave chatbot the ability to handle errors caused by incorrect user input such as missing, incomplete or invalid fields.

Added test cases for A-TextUiTesting.

xhamyo committed:
Implement Level 6

Gave chatbot the ability to delete tasks from the list.

Used collections ArrayList to store Tasks instead of normal Java Array.

Added test cases for delete.

xhamyo committed:
Use Enums

Implement Enums (TODO, DEADLINE, EVENT) for message passing for task types during call to add task.

xhamyo committed:
Implement Level 7

Gave chatbot the ability to save the tasks in the hard disk automatically whenever the task list changes.

Gave chatbot the ability to load the data from the hard disk when the chatbot starts up.

Added toFileFormat method for Tasks for easier file storage.

Added test cases for to test for loading and storing of data.

xhamyo committed:
Update logic

Changed if-else statements to switch to limit magic numbers.

xhamyo committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Hamyo.java

Resolution: change if-else logic to switch, implement loadData() and saveData().

xhamyo committed:
Implement Level 8

Added support for Date in Deadline and Events.

Updated test cases.

xhamyo committed:
Add DateTime Support

Added support for DateTime in Deadline and Events on top of Date.

Updated test cases.

xhamyo committed:
Add listDate

Gave chatbot the ability to list tasks that fall on the specified date using the command "listDate yyyy-mm-dd".

Added occursToday() method in Task to return true if the Task occurs on the specified date and false otherwise.

Added test cases for listDate.

xhamyo committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add listDate
  Add DateTime Support
  Implement Level 8

xhamyo committed:
Use More OOP

Implemented OOP for UI.

Changed Hamyo from static to instance.

xhamyo committed:
Implement Storage

Implemented OOP for Storage.

xhamyo committed:
Implement TaskList

Implemented OOP for TaskList.

Fixed bug where tasks marked in storage outputs extra texts upon chatbot start up.

Updated test cases for tasks marked in storage.

xhamyo committed:
Implement UI

Implemented more OOP for UI. Print statements moved to UI class.

xhamyo committed:
Implement Parser

Implemented OOP for Parser. Added run method to main.

xhamyo committed:
Merge branch 'branch-oop-refactor'

* branch-oop-refactor:
  Implement Parser
  Implement UI
  Implement TaskList
  Implement Storage
  Use More OOP

xhamyo committed:
Implement Packages

Divided classes into packages Hamyo.Misc and Hamyo.Tasks.

Miscellaneous package contains Parser, Storage, UI, HamyoException.

Hamyo package contains Task, Deadline, Event, ToDo, TaskList, TaskType.

xhamyo committed:
Merge branch 'branch-packages'

* branch-packages:
  Implement Packages

xhamyo committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

xhamyo committed:
Implement Unit Tests

Implement JUnit tests for TaskList.

xhamyo committed:
Implement Unit Test for UI

Implemented JUnit Tests for UI.

xhamyo committed:
Merge branch 'branch-junit-tests'

* branch-junit-tests:
  Implement Unit Test for UI
  Implement Unit Tests

xhamyo committed:
Create JAR File

Packaged the Hamyo app as an executable JAR file for ease of distribution.

Updated build.gradle.

xhamyo committed:
Implement Level 9

Gave chatbot the ability to find a task by user input of a keyword.

Added test cases for find command.

xhamyo committed:
Add JavaDoc

Added JavaDoc for all classes and methods.

xhamyo committed:
Tweak Code Style

Tweaked code to comply with coding standard.

xhamyo committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc

xhamyo committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Tweak Code Style

# Conflicts:
#	src/main/java/Hamyo/Misc/HamyoException.java
#	src/main/java/Hamyo/Misc/Parser.java
#	src/main/java/Hamyo/Misc/Storage.java
#	src/main/java/Hamyo/Misc/UI.java
#	src/main/java/Hamyo/Tasks/Deadline.java
#	src/main/java/Hamyo/Tasks/Event.java
#	src/main/java/Hamyo/Tasks/Task.java
#	src/main/java/Hamyo/Tasks/TaskList.java
#	src/main/java/Hamyo/Tasks/ToDo.java

Resolution: Keep JavaDoc from master, Keep formatting from branch.

Other changes:

Tweak test code to comply with coding standards.

Closed scanner when loading tasks via Storage.

xhamyo committed:
UI Bug Fix

Fixed bug caused during merging causing expected and actual printLine() output to differ.

Line length printed by UI.printLine() increased.

Updated test cases.

xhamyo committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Implement Level 9

# Conflicts:
#	src/main/java/Hamyo/Misc/Parser.java
#	src/main/java/Hamyo/Misc/UI.java
#	src/main/java/Hamyo/Tasks/TaskList.java
#	text-ui-test/EXPECTED.TXT

Resolution: Keep new changes from branch-Level-9.

Other changes: Updated test cases with new line length.

rimgik commented on src/main/java/Hamyo.java:
```diff
@@ -0,0 +1,68 @@
+    private final UI ui;
+    private final Storage storage;
+    private final TaskList tasks;
+    private boolean active;
```
I think the name of a boolean variable should be named to sound like a booleans.

rimgik commented on src/main/java/Hamyo/Tasks/Deadline.java:
```diff
@@ -0,0 +1,81 @@
+package Hamyo.Tasks;
```
Shouldn't package name be written in lower case?

rimgik commented on src/main/java/Hamyo/Tasks/Event.java:
```diff
@@ -0,0 +1,99 @@
+     */
+    @Override
+    public boolean occursToday(LocalDate date) {
+        return (date.isAfter(this.fromDate) && date.isBefore(this.toDate)) || date.isEqual(this.fromDate) || date.isEqual(this.toDate);
```
I believe this line is too long. Perhaps we could split it up?

rimgik commented on src/main/java/Hamyo/Tasks/TaskList.java:
```diff
@@ -0,0 +1,198 @@
+     */
+    public void listTasksByKeyword(String str) throws HamyoException {
+        try {
+        if (str.length() <= 1) {
```
I think we are missing indentation here.

rimgik reviewed (commented):
LGTM! Just some coding standard related issues.

xhamyo committed:
Set up Checkstyle for Gradle

xhamyo committed:
Fix Code Style

Fix Java code style to abide by Java coding standard using Checkstyle.

xhamyo committed:
Merge branch 'branch-checkstyle'

* branch-checkstyle:
  Fix Code Style
  Set up Checkstyle for Gradle

xhamyo committed:
Implement GUI

Implemented GUI using JavaFX techonology.

Gave chatbot the ability to echo commands.

xhamyo committed:
Update GUI

Added CSS for styling.

Gave chatbot ability to reply to commands.

xhamyo committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Update GUI
  Implement GUI

xhamyo committed:
Fix Code Style

Checkstyle.

xhamyo committed:
Implement Varargs

Use Varargs for Task, ToDo, Deadline, Event and addTask() for TaskList.

Shifted string logic from TaskList to Parser.

Updated test cases.

xhamyo committed:
Merge branch 'branch-varargs'

* branch-varargs:
  Implement Varargs

xhamyo committed:
Update JavaDocs

Updated JavaDocs with respect to changes made during Implement Varargs.

Changed field names in Task and its subclasses for readability.

charlenetcy commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,58 @@
+
+    private Hamyo hamyo;
+
+    private final Image userIcon = new Image(this.getClass().getResourceAsStream("/images/userIcon.png"));
```
Should they be static final as constants?

charlenetcy commented on src/main/java/Hamyo.java:
```diff
@@ -0,0 +1,103 @@
+     * active, and terminates once this.active is set to false using the bye
+     * command.
+     */
+    public void run() {
```
Maybe refactor the nested try-catch blocks by separating to a method for error handling, ensuring that each block of functionality is more isolated and manageable.

charlenetcy commented on src/main/resources/css/dialog-box.css:
```diff
@@ -0,0 +1,32 @@
+.label {
```
I like how you made multiple sub packages for great readability.

charlenetcy commented on src/main/java/DialogBox.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Flips the dialog box such that the ImageView is on the left and text on the right.
+     */
+    private void flip() {
```
Perhaps modify the method name to be more descriptive?
```suggestion
    private void flip() {
    private void flipDialogBoxOrientation() {
```

charlenetcy reviewed (commented):
Overall, looks great to me!
Very minor violations.
I like how you made it very organised.

xhamyo committed:
Update Varargs

Replaced new String[] with String... missed out during A-Varargs Implementation.

xhamyo committed:
Implement Assertions

Use assertions to ensure correct number of input is given by developer for addTask.

xhamyo committed:
Implement CodeQuality for Misc

Made use of Extract Method to extract methods from Parser.parse into smaller helper functions.

Made use of Extract Method to extract method from constructor and loadData into smaller helper functions.

xhamyo committed:
Implement CodeQuality for Tasks

Made use of Extract Method to extract methods constructor as helper function to setDateTime.

xhamyo committed:
Implement CodeQuality for TaskList

Changed deep nesting to a more readable logic.

xhamyo committed:
Implement Streams

Used Java Streams for listing of tasks.

xhamyo committed:
Merge pull request #1 from xhamyo/branch-A-Assertions

Implement Java Assertions to ensure developer gives the correct number of inputs for addTask.

xhamyo committed:
Merge pull request #2 from xhamyo/branch-A-CodeQuality

Implement Extract Methods for better Code Quality.

xhamyo committed:
Merge branch 'master' into branch-A-Streams

* master:
  Implement CodeQuality for TaskList
  Implement CodeQuality for Tasks
  Implement CodeQuality for Misc
  Implement Assertions

# Conflicts:
#	src/main/java/hamyo/tasks/TaskList.java

xhamyo committed:
Merge pull request #3 from xhamyo/branch-A-Streams

Implement Java Streams for the listing of tasks.

xhamyo committed:
Implement Continuous Integration

Added gradle.yml to github workflows for Continuous Integration.

xhamyo committed:
Merge branch 'branch-ContinuousIntegration'

* branch-ContinuousIntegration:
  Implement Continuous Integration

xhamyo committed:
Implement C-MassOps

Gave parser the ability to process multiple indexes in a single command for mark, unmark and delete.

xhamyo committed:
Remove Duplicates

Remove duplicates from Integer Array caused by input such as delete 1 1 1 1, which might cause buggy behaviour.

xhamyo committed:
GUI Fix and Test Cases

Fix for GUI not displaying information for all mass operations.

Added Test Cases for MassOps.

xhamyo committed:
Fix CheckStyle

Removed unused imports.

xhamyo committed:
Merge branch 'branch-BCD-Extension'

* branch-BCD-Extension:
  Fix CheckStyle
  GUI Fix and Test Cases
  Remove Duplicates
  Implement C-MassOps

xhamyo committed:
Remove IO redirection test

xhamyo committed:
Update UI and GUI

Update UI to store the next String for Terminal and GUI.

Update GUI to close window immediately after bye command.

xhamyo committed:
GUI Visual Bug Fix

Fixed bug where ... and ___ displays on GUI textbox.

xhamyo committed:
Merge branch 'branch-A-BetterGUI'

* branch-A-BetterGUI:
  GUI Visual Bug Fix
  Update UI and GUI

xhamyo committed:
Update UI method name and JavaDocs

BUGGY!

xhamyo committed:
Bug Fix

Squash bug where some Exceptions are not printed due to String overwrite.

xhamyo committed:
Merge branch 'branch-A-BetterGUI'

* branch-A-BetterGUI:
  Bug Fix
  Update UI method name and JavaDocs

xhamyo committed:
Update Test Cases

Update Ui test cases to reset String to empty String for each new test case.

xhamyo committed:
Add Error Handling for invalid dates

Added error handing for invalid Event dates to validate from date is before to date.

Added test cases.

xhamyo committed:
Merge branch 'branch-A-MoreErrorHandling'

* branch-A-MoreErrorHandling:
  Add Error Handling for invalid dates

xhamyo committed:
GUI Test

Conducted final manual testing for GUI.

Squashed bug where InvocationTargetExceptionprivate is thrown for bye command.

Squashed bug where GUI does not close after bye command.

xhamyo committed:
Merge branch 'branch-A-MoreTesting'

* branch-A-MoreTesting:
  GUI Test

xhamyo committed:
Fix GitHub issues

Fixed JavaDocs and boolean naming.

xhamyo committed:
Implement User Guide

Added Ui.png with screenshot of a Hamyo chatbot example execution.

Updated README for User Guide.

xhamyo committed:
Update README

xhamyo committed:
Update README

Fix format issue for Functionalities numbering.

Formate keywords with bold to enhance readability.

xhamyo committed:
Update README
